### DISCUSSION TITLE
train strategy

### DISCUSSION THREAD

**Discussion by yeonhyochoi** [ultralytics/ultralytics#18900] [Q&A]
The train results continued to improve. But now the improvement has stopped. As shown in the document, the background image should be used less than 10%, but I don't think continuing to see the same background during 300 epoch and 500 epoch training will help improve learning. So, we try to replace the training data set every 5 or 10 epochs.

1. All 500,000 object images (with labels) are used for learning.
2. background image 1,500,000 saved in cache with object image(total = 2,000,000)
3. Randomly select 3~10% of object images from the train epoch and use them for the each epoch train.

![image](https://github.com/user-attachments/assets/4ee8a1ed-0901-4d99-b89c-8bc1d805a41d)
![image](https://github.com/user-attachments/assets/6217fdff-3f87-496a-99fb-9a58dbbc70a1)
![image](https://github.com/user-attachments/assets/db83568f-7b86-432a-9c75-179ca38766b6)
![image](https://github.com/user-attachments/assets/1369cb67-eea4-46c4-a578-62eeb516f066)

i also tried like this, but the process very slow. i think it is wrong approach.

![image](https://github.com/user-attachments/assets/11b94443-6f02-4dc2-87a4-9baf32028f8b)
![image](https://github.com/user-attachments/assets/81c7bbfe-bf92-4d5d-abb5-1272e238ec22)

Is there anyone who can give some advice on my code?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yeonhyochoi, thank you for your detailed post and for sharing your training strategy and insights! 🚀 It's great to see how much effort you've put into optimizing your training process and analyzing your results.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for guidance on topics such as dataset balancing and augmentation strategies. For more in-depth information on improving training performance, take a look at our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) 📚. These might help refine your approach further.

If your post is related to a 🐛 Bug Report, we kindly ask for a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can assist you quickly and efficiently.

If you’re looking for additional feedback on your custom training ❓, providing more details, such as logs, sample images (if possible), or snippets of your dataset metadata could help facilitate better advice from both the community and our team.

In terms of slower processing speed with your approach, you may also want to explore integrating dataset sampling techniques or caching optimizations. Check the [Dataset Guide](https://docs.ultralytics.com/datasets/) for tips on efficiently handling large datasets (such as your 2,000,000-image dataset). 

For **real-time help**, join the discussion on our [Discord](https://discord.com/invite/ultralytics) server 🎧. If you prefer topic-specific threads or detailed posts, our [Discourse Community](https://community.ultralytics.com/) or [Subreddit](https://reddit.com/r/Ultralytics) would be excellent places to collaborate with other developers.

## Upgrade
First, please ensure you're running the latest version of `ultralytics` for compatibility and performance enhancements. Update it in a [**Python>=3.8**](https://www.python.org/) environment with:

```bash
pip install -U ultralytics
```

Make sure all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) are installed. This will ensure that you're leveraging the latest improvements for data handling, training, and optimization.

## Environments
Consider running your experiments in one of the following verified environments, which include preinstalled dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn):
- Cloud Notebooks: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>, <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>, or <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/), or [Docker](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) setups <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

---

This is an automated response 🤖, but an Ultralytics engineer will review your discussion and provide further assistance shortly. Thank you for being a part of the Ultralytics community! 🌟

**Comment by glenn-jocher** [comment]
@yeonhyochoi it seems you're experimenting with data strategies to improve training outcomes. While replacing or dynamically sampling datasets can be beneficial in some cases, it introduces complexity and can slow down training significantly, as you've observed. Instead, consider techniques like **data augmentation**, **multi-scale training**, or **mosaic augmentation**, which can increase data variety without the need for dataset replacement. For background images, maintaining a 3–10% ratio is a sound practice, provided they are representative of your deployment environment. If you need further optimization, you can explore hyperparameter tuning or early stopping when metrics plateau to save resources. For more details, refer to the training tips [here](https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/).
